swarm/src/lib.rs

```diff
@@ -1231,21 +1235,21 @@ where
 pub struct SwarmPollParameters<'a> {
     local_peer_id: &'a PeerId,
     supported_protocols: &'a [Vec<u8>],
-    listened_addrs: &'a [Multiaddr],
+    listened_addrs: Vec<&'a Multiaddr>,
     external_addrs: &'a Addresses,
 }
 
 impl<'a> PollParameters for SwarmPollParameters<'a> {
     type SupportedProtocolsIter = std::iter::Cloned<std::slice::Iter<'a, std::vec::Vec<u8>>>;
-    type ListenedAddressesIter = std::iter::Cloned<std::slice::Iter<'a, Multiaddr>>;
+    type ListenedAddressesIter = std::iter::Cloned<std::vec::IntoIter<&'a Multiaddr>>;
     type ExternalAddressesIter = AddressIntoIter;
 
     fn supported_protocols(&self) -> Self::SupportedProtocolsIter {
         self.supported_protocols.iter().cloned()
     }
 
     fn listened_addresses(&self) -> Self::ListenedAddressesIter {
-        self.listened_addrs.iter().cloned()
+        self.listened_addrs.clone().into_iter().cloned()
```

Comment:
```suggestion
        self.listened_addrs.clone().into_iter()
```

Is this needed? Isn't it already an owned iterator?

Reply:
`self.listened_address` is an `Vec<&'a Multiaddr>`, but we need `ExactSizeIterator<Item = Multiaddr>`.
So the first clone is to make at an owned iterator, the second `cloned` is for `Item = &Multiaddr` -> `Item = Multiaddr`.

In the end what we need is to create a `ExactSizeIterator<Item = Multiaddr>` from a `HashMap<ListenerId, SmalLVec<Multiaddr>>`. The current implementation is the only one I was able to come up with, but I'd be grateful for any ideas of how to do this cleaner.

Reply:
Thanks for detailing this! You are right.

> The current implementation is the only one I was able to come up with, but I'd be grateful for any ideas of how to do this cleaner.

Played around with this as well. Couldn't come up with something better. Also, performance wise, I don't think it matters. And long term I would like to get rid of `PollParameters`, though I don't see this happening any time soon.